File: nvim/lua/rw-ai/frames.lua
-- Length-prefixed JSON framing — the Lua mirror of src/daemon/frames.ts.
-- A frame is a 4-byte big-endian uint32 byte length, then that many UTF-8 bytes
-- of JSON. LuaJIT (what Neovim embeds) has no string.pack, so the header is
-- packed/unpacked by hand.

local M = {}

local function u32_be(n)
  return string.char(
    math.floor(n / 16777216) % 256,
    math.floor(n / 65536) % 256,
    math.floor(n / 256) % 256,
    n % 256
  )
end

function M.encode(value)
  local json = vim.json.encode(value)
  return u32_be(#json) .. json
end

-- Stateful decoder: push raw chunks, get back every complete decoded value.
-- A partial trailing frame is retained for the next push. A corrupt (non-JSON)
-- payload is skipped without desyncing the stream (the length prefix keeps it
-- aligned).
function M.new_decoder()
  local buffer = ''
  return {
    push = function(chunk)
      buffer = buffer .. chunk
      local out = {}
      while #buffer >= 4 do
        local a, b, c, d = buffer:byte(1, 4)
        local length = ((a * 256 + b) * 256 + c) * 256 + d
        if #buffer < 4 + length then
          break
        end
        local payload = buffer:sub(5, 4 + length)
        buffer = buffer:sub(5 + length)
        local ok, decoded = pcall(vim.json.decode, payload)
        if ok then
          out[#out + 1] = decoded
        end
      end
      return out
    end,
  }
end

return M

File: src/daemon/frames.ts
// Length-prefixed JSON framing for the observer daemon's pipe/socket transport:
// a 4-byte big-endian uint32 byte length, then that many UTF-8 bytes of JSON.
// A stream carries many frames; the decoder buffers partial reads and yields
// whole values only.

const HEADER_BYTES = 4;
// 8 MiB default cap: a status snapshot is tiny, so any larger declared length is
// a corrupt/hostile stream — reject it rather than buffer unboundedly (anti-OOM).
const DEFAULT_MAX_FRAME_BYTES = 8 * 1024 * 1024;

export const encodeFrame = (value: unknown): Buffer => {
  const json = Buffer.from(JSON.stringify(value), 'utf8');
  const header = Buffer.allocUnsafe(HEADER_BYTES);
  header.writeUInt32BE(json.length, 0);
  return Buffer.concat([header, json]);
};

export type FrameDecoder = {
  // Feed raw bytes; returns every complete frame that became available. A partial
  // trailing frame is retained for the next push. Throws on an over-cap length or
  // a non-JSON payload — the caller (the server) drops that connection.
  readonly push: (chunk: Buffer) => unknown[];
};

export const createFrameDecoder = (
  maxFrameBytes: number = DEFAULT_MAX_FRAME_BYTES,
): FrameDecoder => {
  // Annotated as the wide Buffer type: Buffer.alloc yields Buffer<ArrayBuffer>
  // while Buffer.concat yields Buffer<ArrayBufferLike>, and we reassign with both.
  let buffered: Buffer = Buffer.alloc(0);
  return {
    push(chunk) {
      buffered = buffered.length === 0 ? chunk : Buffer.concat([buffered, chunk]);
      const out: unknown[] = [];
      while (buffered.length >= HEADER_BYTES) {
        const length = buffered.readUInt32BE(0);
        if (length > maxFrameBytes) {
          throw new Error(`frame too large: ${length} bytes exceeds cap ${maxFrameBytes}`);
        }
        if (buffered.length < HEADER_BYTES + length) {
          break; // frame not fully arrived yet
        }
        const payload = buffered.subarray(HEADER_BYTES, HEADER_BYTES + length);
        out.push(JSON.parse(payload.toString('utf8')));
        buffered = buffered.subarray(HEADER_BYTES + length);
      }
      return out;
    },
  };
};
